File: ai-learning-cpp/src/domain/knowledge/relation.cpp
/**
 * @file relation.cpp
 * @brief 知识关系实现
 */

#include "ai_learning/domain/knowledge/relation.hpp"

#include <algorithm>

namespace ai_learning::domain::knowledge {

Relation::Relation(std::string source_id, std::string target_id,
                   std::string type, double confidence,
                   int evidence_count,
                   std::map<std::string, std::string> metadata)
    : source_id_(std::move(source_id)),
      target_id_(std::move(target_id)),
      type_(std::move(type)),
      confidence_(std::max(0.0, std::min(1.0, confidence))),
      evidence_count_(evidence_count),
      metadata_(std::move(metadata)) {}

auto Relation::strengthen(double delta) const -> Relation {
    return Relation(source_id_, target_id_, type_,
                    std::min(1.0, confidence_ + delta),
                    evidence_count_ + 1, metadata_);
}

auto Relation::weaken(double delta) const -> Relation {
    return Relation(source_id_, target_id_, type_,
                    std::max(0.0, confidence_ - delta),
                    evidence_count_, metadata_);
}

auto Relation::to_map() const -> std::map<std::string, std::string> {
    auto result = metadata_;
    result["_source_id"]     = source_id_;
    result["_target_id"]     = target_id_;
    result["_type"]          = type_;
    result["_confidence"]    = std::to_string(confidence_);
    result["_evidence_count"] = std::to_string(evidence_count_);
    return result;
}

auto Relation::from_map(const std::map<std::string, std::string>& m)
    -> Relation {
    return Relation(
        m.count("_source_id") ? m.at("_source_id") : "",
        m.count("_target_id") ? m.at("_target_id") : "",
        m.count("_type") ? m.at("_type") : "",
        m.count("_confidence") ? std::stod(m.at("_confidence")) : 1.0,
        m.count("_evidence_count")
            ? std::stoi(m.at("_evidence_count")) : 0,
        {}  // metadata（简化，不反序列化内部 metadata）
    );
}

}  // namespace ai_learning::domain::knowledge

File: archive/parallel-learning/src/knowledge/relation.py
"""知识关系 — 知识图谱中的边"""

from dataclasses import dataclass, field
from typing import Any, Dict, List


# 关系类型常量
REL_IS_A = 'is_a'
REL_HAS_PROPERTY = 'has_property'
REL_PART_OF = 'part_of'
REL_CAUSES = 'causes'
REL_SIMILAR_TO = 'similar_to'
REL_COLLOCATES = 'collocates_with'
REL_RELATED = 'related_to'


@dataclass
class Relation:
    """知识图谱中的关系（边）

    连接两个实体，表达它们之间的语义关系。
    """
    source_id: str
    target_id: str
    type: str
    confidence: float = 1.0
    evidence_count: int = 0
    metadata: Dict[str, Any] = field(default_factory=dict)

    def strengthen(self, delta: float = 0.1) -> None:
        """加强关系置信度"""
        self.confidence = min(1.0, self.confidence + delta)
        self.evidence_count += 1

    def weaken(self, delta: float = 0.1) -> None:
        """削弱关系置信度"""
        self.confidence = max(0.0, self.confidence - delta)

    def to_dict(self) -> dict:
        return {
            'source_id': self.source_id,
            'target_id': self.target_id,
            'type': self.type,
            'confidence': self.confidence,
            'evidence_count': self.evidence_count,
            'metadata': self.metadata,
        }

    @classmethod
    def from_dict(cls, d: dict) -> 'Relation':
        return cls(
            source_id=d['source_id'],
            target_id=d['target_id'],
            type=d['type'],
            confidence=d.get('confidence', 1.0),
            evidence_count=d.get('evidence_count', 0),
            metadata=d.get('metadata', {}),
        )
